### DISCUSSION TITLE
Error loading `pyodide.asm.js` in a React App

### DISCUSSION THREAD

**Discussion by franknmungai** [pyodide/pyodide#3032] [Q&A]
Hello, I'm having an issue trying to load up `pyodide.asm.js` when calling `loadPyodide()`. I have already hosted the `pyodide.asm.js` file in my React app in the public folder under the following URL path: `public/static/js/pyodide.asm.js`

I keep getting the following error, even though I'm serving up the file in the correct directory.

```
pyodide|lazy|groupOptions: {}|strict namespace object:5 Uncaught (in promise) Error: Cannot find module '/static/js/pyodide.asm.js'
    at pyodide|lazy|groupOptions: {}|strict namespace object:5:1
    at async loadScript (compat.ts:134:1)
    at async loadPyodide (pyodide.ts:290:1)

```
I got the `pyodide.asm.js ` file from the pyodide-build-0.21.1 in the releases section after extracting the tar file. How can I resolve this error and load up `pyodide.asm.js` correctly?
Here are some of my library versions:

```
 "pyodide": "^0.21.1",
"react": "^17.0.1"
"react-dom": "^17.0.1",
"react-scripts": "5.0.1",
```

**Comment by hoodmane** [comment]
Try passing the full URL to `pyodide.asm.js` as `indexURL` as an option for `loadPyodide`.

**Reply by franknmungai** [reply]
Hello @hoodmane,
Yes, I've been trying to do that... I've tried adding the full URL path to where the file is being served i.e., 
`loadPyodide({ indexURL: '/static/js/' })` and `loadPyodide({ indexURL: 'http://localhost:3000/static/js/' })` , which didn't work.
I've also tried to add a relative path to the folder containing the `pyodide.asm.js`  file within my code, i.e., `loadPyodide({ indexURL: '../pyodide' })` but that also didn't work out. I've also tried to add the `pyodide.asm.js` in a custom folder within the `node_modules `and then used the `require.resolve` method to resolve the file name i.e.,  `loadPyodide({ indexURL: require.resolve("pyodide-wasm") }) ` however, I still got the same  _"module not found error"_

**Reply by hoodmane** [reply]
Could you provide a minimal example of the problem?

**Reply by franknmungai** [reply]
Here's a minimal example that includes a React component that tries to call the `loadPyodide` function when first mounted.

```
import { loadPyodide, PyodideInterface } from 'pyodide';
import { useEffect, useRef } from 'react';

export const App = () => {
  const pyodideRef = useRef<PyodideInterface>();

  useEffect(() => {
    (async () => {
      // Throws a "module not found" error
      pyodideRef.current = await loadPyodide({
        indexURL: '/static/js',

        // Some other alternatives I've tried
        // indexURL: '../../pyodide' -> relative import to folder containing pyodide.wasm.js
        //indexURL: 'require.resolve('pyodide-wasm')' -> Trying resolve the filename 
      });
    })();
  });
  return <div></div>;
};
```

**Reply by hoodmane** [reply]
I don't really know much about react, how do I run this? Can you provide also an appropriate `package.json` or `npm install` command and the build command?

**Reply by franknmungai** [reply]
Hi @hoodmane, 
Here's a link to a repository for a simple react app that reproduces the issue:

[Repo](https://github.com/franknmungai/py-debug)

You can try cloning it and reviewing the code. The `pyodide.asm.js` file is in the public folder under `public/static/js`, and the application source code is in the `src` folder. The main file is the `App.js`. To run the project, just install the packages by running `npm install ` in the project root folder, and then running `npm start`.
The react app should open on port 3000 -> [http://localhost:3000](http://localhost:3000)

Open the browser dev tools and you'll see the pyodide load error.
```
pyodide|lazy|groupOptions: {}|strict namespace object:5 Uncaught (in promise) Error: Cannot find module 'http://localhost:3000/static/js/pyodide.asm.js'
    at pyodide|lazy|groupOptions: {}|strict namespace object:5:1
    at async loadScript (pyodide.mjs:1:1)
    at async loadPyodide (pyodide.mjs:1:1)
    at async App.js:10:1
```

**Comment by Amith225** [comment]
i am facing the exact issue, did you find any solution?

**Comment by Amith225** [comment]
![image](https://user-images.githubusercontent.com/75326634/190335955-e346ec7e-a599-4691-8642-3fd09abf1cfa.png)
its weird, ```loadPyodide``` won't work for https://cdn.jsdelivr.net/pyodide/v0.21.2/full either

**Comment by ryanking13** [comment]
It seems like the problem is that [Webpack does some static analyse on imports at build time](https://github.com/webpack/webpack/issues/6680) which breaks dynamic imports on URLs.

I think #3094 was a fix for this problem (prepending `/* webpackIgnore: true */` on every import command), but the fix was only applied  to `pyodide.js` not in `pyodide.mjs` so the error still happens in modules.

@hoodmane

**Comment by ryanking13** [comment]
I think this is fixed with #3132 in the dev url (` https://cdn.jsdelivr.net/pyodide/dev/full`). Could you check if it is fixed?
